python/ray/serve/tests/test_controller_recovery.py

```diff
@@ -291,13 +291,116 @@ def check_replica(replica_state=None):
     # Make sure the replica is in STOPPING state.
     wait_for_condition(lambda: len(check_replica(ReplicaState.STOPPING)) > 0)
 
-    # Unblock the request and the replica will be stopped.
-    signal.send.remote()
+    # The graceful shutdown timeout of 3 seconds should be used
+    wait_for_condition(lambda: len(check_replica()) == 0, timeout=5)
+    # Application should be removed soon after
     wait_for_condition(
         lambda: serve_instance.get_serve_status("app").app_status.status
-        == ApplicationStatus.NOT_STARTED
+        == ApplicationStatus.NOT_STARTED,
+        timeout=1,
+    )
+
+
+def test_recover_deleting_deployment(serve_instance):
+    """Test that replicas that are stuck on __del__ when the controller crashes,
+    is properly recovered when the controller is recovered.
+
+    This is similar to the test test_replica_deletion_after_controller_recover,
+    except what's blocking the deployment is __del__ instead of ongoing requests
+    """
+    # TODO(zcin): Update this test to use serve.run() and application after the deploy
+    # refactor adds checkpointing for applications. Right now we checkpoint the user
+    # submitted config and use that for recovery, so applications that are in the
+    # process of deleting when the controller crashes, will not be recovered when the
+    # controller recovers (although its deployment will be recovered). This should be
+    # fixed after we refactor the application state manager.
+
+    signal = SignalActor.remote()
+
+    @serve.deployment
+    class A:
+        async def __del__(self):
+            await signal.wait.remote()
+
+    A.deploy()
+
+    @ray.remote
+    def delete_task():
+        serve.get_deployment("A").delete()
+
+    # Delete deployment and make sure it is stuck on deleting
+    delete_ref = delete_task.remote()
+    print("Started task to delete deployment `A`")
+
+    def deployment_deleting():
+        check = True
+        # Confirm deployment is in updating state
+        status = serve_instance.get_all_deployment_statuses()[0]
+        check |= status.name == "A" and status.status == "UPDATING"
+
+        # Confirm replica is stopping
+        replicas = ray.get(
+            serve_instance._controller._dump_replica_states_for_testing.remote("A")
+        )
+        assert replicas.count(states=[ReplicaState.STOPPING]) == 1
```

Comment:
I'm not sure if those replicas will immediately become stopping state when delete is called. But if not, this assertion will get triggered and fail the test. Not sure if it's better to wait for this to happen (by doing `return replicas.count(states=[ReplicaState.STOPPING]) == 1` instead), but feel free to keep as is if it's intentional

Reply:
Good point! I believe `wait_for_condition` handles exceptions and treats it as if the function returned false, but there's no reason to use an assert here instead of returning false - I'll change it